### ISSUE TITLE
Workspace switcher: with HighContrast theme, no way to determine the current workspace

### ISSUE THREAD

**Issue by alexarnaud** [mate-desktop/mate-panel#807]
#### Expected behaviour
Current desktop should be distinguishable by a visual effect

#### Actual behaviour
There is no way to determine the currently selected workspace.

![capture du 2018-05-09 14-33-40](https://user-images.githubusercontent.com/4191500/39814803-0aebfdce-5396-11e8-9aae-52df641e2844.png)

#### Steps to reproduce the behaviour
1) Select the GNOME High Contrast theme
2) Open Pluma on the workspace 1, maximize the Pluma window
3) Open Mate terminal on the workspace 2, maximize the Mate terminal window

#### MATE general version
1.20.1

#### Package version
1.20.1-3

#### Linux Distribution
Debian Sid

**Comment by alexarnaud** [comment]
@vkareh clearly. I don't see the color change (I'm visual-impaired and I have difficulties to see little color change)

 See a screenshot from Debian 8 "Jessie" with Mate 1.8 GTK2
![capture du 2018-05-09 18-24-16](https://user-images.githubusercontent.com/4191500/39826810-3a2f327a-53b6-11e8-8075-fc67f2f23b52.png)

**Comment by vkareh** [comment]
@alexarnaud - I'm not able to reproduce this. I'm on a fresh Debian Sid with the High Contrast theme from the `gnome-accessibility-themes` package and this is how it looks on my side:

![screenshot at 2018-05-30 06-49-11](https://user-images.githubusercontent.com/259780/40716187-1bece566-63d6-11e8-99be-f5b6d9ca92d0.png)

Your screenshot shows pale blue for your selected workspace 1, and slightly darker blue for the rest. My screenshot shows pale blue for my selected workspace 2, and gray for the rest.

This theme selects the Adwaita icons by default, so I tried the HighContrast icons and still has the same colors on my side (but now the icons look more like your screenshot from Debian Jessie).

Analyzing your original screenshot, it looks like you have a full-screen window on the last workspace, and maximized windows on the other ones. Even trying the same setup, I still get the colors on my screenshot. Honestly it looks more like a bug in how the windows are being marked as selected, rather than a theme issue. What is your window manager?

Also, do you have any customizations on your theme? Or is it possible that the theme files have been updated since and it's no longer an issue?

**Comment by alexarnaud** [comment]
Le 30/05/2018 à 13:02, Victor Kareh a écrit :
> screenshot at 2018-05-30 06-49-11 
> <https://user-images.githubusercontent.com/259780/40716187-1bece566-63d6-11e8-99be-f5b6d9ca92d0.png>
> 

Keep in mind I'm visual-impaired and most of low-vision people have 
difficulties to see colors.

To be honest, with your screenshot I can't tell you what is the selected 
workspace. For me the issue still appears, with High Contrast theme, we 
should have something really easy to see like it is on GTK2.

Could you please tell me if for you GTK2 screenshot is more contrasted ?

 > Also, do you have any customizations on your theme? Or is it possible
 > that the theme files have been updated since and it's no longer an issue?

I don't think so. Sometimes I'm running redshift-gtk and the screenshot 
I've been made on a VM with Debian Sid.

Best regards,
Alex.

**Comment by lukefromdc** [comment]
Your screenshot of MATE 1.8 is an older, GTK2 version of MATE, and GTK3 is now used. Thus, if you were using the HighContrast theme, that changes (like all themes did) from the GTK2 version to the GTK3 version.

Here's what it looks like on GTK 3.22.30 with mate-panel 1.21 from git master as of today:


In current git master, the HighContrast and HichContrastInverse themes come directly from GTK and are not our themes We used to split out a separate ContrastHighInverse based on the GNOME theme but not anymore as differences had declined to near zero.. 

These themes are included in GTK3 itself and are compiled into it whether than being in a theme directory. They use the same CSS and inverse mode just switches the colors black for white for the most part. Bugs in them are thus bugs in the GTK3 package and need to be reported there if they are general bugs.

Specific problems in a MATE package such as mate-panel when run with a GNOME theme are a little harder to fix, as they require css in the package itself, cannot conflict with or override other themes, and thus potentially can require a runtime check to get the theme name.

**Comment by vkareh** [comment]
Okay, this is starting to make sense to me. So, in my screenshot the colors are different (gray vs blue), but the contrast ratio is 1.07. In your screenshot the colors are similar (blue vs blue), but the contrast is actually higher: 1.48.

In your Debian Jessie screenshot (gray vs black) the contrast level is 5.24. And WCAG 2.0 guidelines specify a contrast level of at least 4.5, if I understand this correctly.

I found a relevant commit: https://github.com/mate-desktop/mate-panel/commit/956d36c0 that adds a fall-back color in case the theme does not provide one. That's the color your screenshot shows.

I checked and it seems like the High Contrast theme does not provide a color for this. The best solution would be to update that theme. Alternatively, we could change the fallback color so that it defaults to something that works with the High Contrast theme (and any other themes can just override it, as they already do).

Yet another possibility would be for the applet to query the current theme data and grab an appropriate color from there, then use that as the fallback, thus forcing the correct theme color even when it's not set. I'll play with this some more in the coming days.

**Comment by lukefromdc** [comment]
We cannot update the HighContrast theme, that comes from GNOME and is compiled directly into GTK itself

**Comment by alexarnaud** [comment]
Le 30/05/2018 à 22:53, Victor Kareh a écrit :
> Okay, this is starting to make sense to me. So, in my screenshot the 
> colors are different (gray vs blue), but the contrast ratio is 1.07. In 
> your screenshot the colors are similar (blue vs blue), but the contrast 
> is actually higher: 1.48.
> 
> In your Debian Jessie screenshot (gray vs black) the contrast level is 
> 5.24. And WCAG 2.0 guidelines specify a contrast level of at least 4.5, 
> if I understand this correctly.

What you're telling seems relevant to me. How do you determine the 
contrast level?

> I checked and it seems like the High Contrast theme does not provide a 
> color for this. The best solution would be to update that theme. 
> Alternatively, we could change the fallback color so that it defaults to 
> something that works with the High Contrast theme (and any other themes 
> can just override it, as they already do).
> 
> Yet another possibility would be for the applet to query the current 
> theme data and grab an appropriate color from there, then use that as 
> the fallback, thus forcing the correct theme color even when it's not 
> set. I'll play with this some more in the coming days.

It seems it's possible to do some changes on GTK3 to increase Mate 
support: https://gitlab.gnome.org/GNOME/gtk/merge_requests/125

Unfortunately, GTK3 guys wouldn't do something on this area until we 
propose a merge request.

Best regards,
Alex.

**Comment by vkareh** [comment]
hmmm... I posted a comment earlier today and it doesn't show up here... :exclamation::question:

@alexarnaud - I use this website: http://contrast-ratio.com

Also, I can try to submit a fix for GTK, but that might take more time (and it's a toss-up as to whether they'll accept it). In the meantime I submitted a fix for the mate-panel (#813) that essentially queries the current theme for the most appropriate color and sets that as the fall-back.

**Comment by muktupavels** [comment]
If it will be application specific change / fix then I am pretty sure that they will not accept, at least I would not. If that is some generic fix that works for all apps then most likely it will be accepted. :)

For GTK builtin themes we have (should?) to provide our own extra style if our apps requires it. You can detect when user changes theme, you can get current theme name and load/unload extra style as needed. Check what GNOME apps do:
- https://gitlab.gnome.org/GNOME/nautilus/blob/master/src/nautilus-application.c#L1136
- https://gitlab.gnome.org/GNOME/gnome-panel/blob/master/gnome-panel/main.c#L43
- https://gitlab.gnome.org/GNOME/gedit/blob/master/gedit/gedit-app.c#L679

@lukefromdc it is not about mate-panel and/or gnome-panel... Do you really expect GTK to start ship application specific css? I don't... As you see they don't add nautilus or any other application specific css.

I think GTK could add API or something that would allow easy way for apps to extend builtin themes, but then how often apps need custom css...

P.S. @lukefromdc about what/which network-manager-applet patch you speak? where did you submit it? link to bug report?

**Comment by muktupavels** [comment]
@lukefromdc I guess I found... https://github.com/GNOME/network-manager-applet/pull/5, right?

Did you know that it is mirror and developers does not look there? There are chances that developer/developers even does not know that it is mirrored in GitHub. If you want to contribute then you should do your homework - find where projects hosts code, where bugs should be filled and not post something in random place on internet...

At that time it was cgit at git.gnome.org for code and bugzilla at bugzilla.gnome.org for bug reports. That is where you should have reported bug and/or posted patch. Or maybe sending to mailing list also would have been good enough, you can find it in https://wiki.gnome.org/Projects/NetworkManager.

Now you probably should do it here - https://gitlab.gnome.org/GNOME/network-manager-applet.

**Comment by lukefromdc** [comment]
An old account I had there no longer works and I was unable to make another. With my radically changed browser settings a LOT of websites do not work, as I have chosen to prioritize anti-tracking and anti-ads over functionality

**Comment by lukefromdc** [comment]
In my network-manager-applet repo,
https://github.com/lukefromdc/network-manager-applet/commit/47e1373c37a11b84dfa6d29f188ba80a45db04df
fixes the lock icon not being shown on an encrypted connection, no idea if that was later fixed upstream at the moment.
https://github.com/lukefromdc/network-manager-applet/commit/d67657ad5cbea53a5499e41cc30bb6129eef1a8d
Applies the same support for mate-panel themes differing from the main GTK theme as in mate-panel and mate-applet. That one is probably not of interest to GNOME
